### ISSUE TITLE
Custom Ops + torch.compile bug/clarification

### ISSUE THREAD

**Issue by alihassanijr** [pytorch/pytorch#137437]
### 🐛 Describe the bug

I'm trying to register a custom CPP op with torch.library instead of the "old" way of directly calling the dlopened handle from an autograd function. I followed the [custom ops tutorial](https://pytorch.org/tutorials/advanced/python_custom_ops.html).

The issue I'm facing is that even though I'm calling `.contiguous()` on inputs to the custom op, the generated graph will have "incorrect" strides, and the contiguous checks in the op will end up failing.

Input tensors are of shape `[2, 4, 7, 7, 32]`, and with pytorch's contiguous memory layout, that should translate into stride of `[6272, 1568, 224, 32, 1]`. However, the generated graph will produce those buffers with stride of `[6400, 1600, 224, 32, 1]`.

Here is a minimum reproducible script that does not include any dependencies, or the custom CPP op. I'm hoping I can be pointed in the right direction as to what I'm doing wrong:

https://gist.github.com/alihassanijr/f211960a122dbf5c182b72e9e57aa3d9

If I had to guess, I'd say the GEMM preceding the `.contiguous()` calls and the custom op is padding the input for memory alignment, and somewhere something's breaking, but even if that assumption is true, it doesn't really give me any ideas on what to try next.

A no brainer is that I could call `.contiguous()` on the tensors inside the custom op, but the I don't know how to write a backwards pass for them. Doing so will obviously go around the runtime issue, but if I train models with it, there's a very noticeable error.

I tried using `clone` and `copy`, but the strides emitted in the final graph are still wrong.

I've tried this with both torch `2.4.0` and `2.4.1`, both with CTK 12.4 and on an A100-SXM.

### Versions

Collecting environment information...
PyTorch version: 2.4.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.25.3
Libc version: glibc-2.35

Python version: 3.10.15 (main, Oct  3 2024, 07:27:34) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-118-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               256
On-line CPU(s) list:                  0-254
Off-line CPU(s) list:                 255
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7763 64-Core Processor
CPU family:                           25
Model:                                1
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            2
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3529.0520
CPU min MHz:                          0.0000
BogoMIPS:                             4900.35
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                       AMD-V
L1d cache:                            4 MiB (128 instances)
L1i cache:                            4 MiB (128 instances)
L2 cache:                             64 MiB (128 instances)
L3 cache:                             512 MiB (16 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-63,128-191
NUMA node1 CPU(s):                    64-127,192-254
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] torch==2.4.0+cu124
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.19.0+cu124
[pip3] triton==3.0.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] torch                     2.4.0+cu124              pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.19.0+cu124             pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @zou3519 @bdhirsh

**Comment by eellison** [comment]
cc @zou3519

**Comment by zou3519** [comment]
>  However, the generated graph will produce those buffers with stride of [6400, 1600, 224, 32, 1].

Which graph are you looking at to determine this?

**Comment by alihassanijr** [comment]
I can attach the graph directly, but the [script attached above](https://gist.github.com/alihassanijr/f211960a122dbf5c182b72e9e57aa3d9) it will generate it as well.

**Comment by zou3519** [comment]
@eellison this is an example where we need exact strides. Also I learned today that the requires_stride_order doesn't actually mean dense strides

**Comment by zou3519** [comment]
@alihassanijr for a workaround, you can call .contiguous() inside the kernel itself. What are the problems you have with computing backwards? The .contiguous() call shouldn't change the gradient formula.

**Comment by alihassanijr** [comment]
@zou3519 I've tried that, and it does go around the runtime error, but backprop is not functionally correct. Meaning if I train a model with the custom op, I'm not able to reproduce the results.

**Comment by alihassanijr** [comment]
I'll try and do that. But I want to clarify something first. Should both the forward and backward kernels call `.contiguous()` on input tensors, or is it possible to do that once in `setup_context` instead?

I think in the old way of creating autograd functions, `setup_context` was essentially part of forward, so it looked like forward pass just makes everything contiguous, and stores the contiguous version of the backward pass, meaning backward pass wouldn't have to do it again. I'm not sure whether that's what actually happens, but I just want to make sure I'm not doing it in a wrong or inefficient way.

**Comment by zou3519** [comment]
> Should both the forward and backward kernels call .contiguous() on input tensors, or is it possible to do that once in setup_context instead?

Both of the kernels  (custom_op and custom_op_backward) need to call .contiguous() -- torch.compile/Inductor is able to optimize the inputs to `custom_op` and `custom_op_backward` and change their striding

**Comment by zou3519** [comment]
> I.e. is there an equivalent for [custom_fwd](https://pytorch.org/docs/stable/notes/amp_examples.html#autocast-and-custom-autograd-functions) and [custom_bwd](https://pytorch.org/docs/stable/notes/amp_examples.html#autocast-and-custom-autograd-functions) for torch.library ops?

Try https://gist.github.com/zou3519/7f18ed56888053aca4f7ae53dd710414, we're working on adding a more first-class API for this.

**Comment by alihassanijr** [comment]
Thank you! Is there a way to have it support both bfloat16 and float16 though, depending on what the amp context dtype is?

**Comment by exclamaforte** [comment]
Closing because #137979 seems like the followup. @alihassanijr feel free to reopen with repro of NaNs in backwards if it's still an issue.
